### PR TITLE
Feature: Install official plugin

### PR DESCRIPTION
**Pull request by justinehell**
Add prompt to choose between official or custom plugins via `install` command.
Resolve https://github.com/ditrit/leto-modelizer-plugin-cli/issues/8

### CODE REVIEW

File: src/commands/Install.js [hunk 1]
```diff
@@ -1,70 +1,71 @@
...
+        if (plugins.length === 0) {
+          return;
+        }
+
+        plugins.forEach((plugin) => {
+          pluginNames.push(plugin.name);
+          repositoryUrls.push(plugin.url);
+        });
+      }
+      console.log(`\n${chalk.blue.bold('⚒')} Installing plugin(s) via npm...`);
```
Comment (k-y0u) [inline src/commands/Install.js:36]: jump line before

File: src/services/helper.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+async function onePluginToInstall(plugins) {
+  console.log(`\n${chalk.italic(' One plugin available')}\n`);
+
+  const { name, url, version } = plugins[0];
+  const confirmInstall = (await pluginPrompt.confirmInstallation({ name, version })).value;
+
+  if (!confirmInstall) {
+    console.log(`\n${chalk.red('✘')} No plugin has been installed.`);
+    return [];
```
Comment (k-y0u) [inline src/services/helper.js:17]: jump line before

File: src/services/helper.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  return [{ name, url }];
+}
+
+async function multiPluginsToInstall(plugins) {
+  const pluginNames = plugins.map(({ name }) => name);
+  console.log(`\n${chalk.blue.bold('⚒')} Official plugin(s) : ${pluginNames.join(', ')}.\n`);
```
Comment (k-y0u) [inline src/services/helper.js:25]: jump line before

File: src/services/helper.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+  const pluginList = plugins.map((plugin) => ({ title: `${plugin.name} (${plugin.version})`, value: plugin.url }));
+  const selectedPlugins = (await pluginPrompt.getOfficialPlugins(pluginList)).type;
+
+  if (selectedPlugins.length === 0) {
+    console.log(`\n${chalk.red('✘')} No plugin has been installed.`);
+    return [];
```
Comment (k-y0u) [inline src/services/helper.js:32]: jump line before

File: src/services/helper.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  return selectedPlugins
+    .map((url) => ({
+      name: plugins.find((plugin) => plugin.url === url).name,
+      url,
+    }));
+}
+
+function parseArguments(args) {
+  const parsedArgs = {};
+  args.forEach((arg) => {
```
Comment (k-y0u) [inline src/services/helper.js:44]: jump line before

File: src/services/helper.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    const [name, value] = arg.split('=');
+    parsedArgs[name] = value;
+  });
+  return parsedArgs;
```
Comment (k-y0u) [inline src/services/helper.js:48]: jump line before

File: src/services/helper.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+      console.log(`\n${chalk.yellow('⚠')} No official plugin to install.`);
+    } else if (plugins.length === 1) {
+      selectedPlugins = await onePluginToInstall(plugins);
+    } else {
+      selectedPlugins = await multiPluginsToInstall(plugins);
+    }
+  } else {
+    console.log(`\n${chalk.yellow('⚠')} No official plugin to install.`);
+  }
+  return selectedPlugins;
```
Comment (k-y0u) [inline src/services/helper.js:67]: jump line before

File: src/services/helper.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+async function getCustomPlugin() {
+  customPlugin.name = (await pluginPrompt.getName()).pluginName;
+  customPlugin.url = (await pluginPrompt.getRepositoryUrl()).repositoryUrl;
+  return customPlugin;
```
Comment (k-y0u) [inline src/services/helper.js:73]: jump line before

File: src/services/helper.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+exports.rules = { pluginNameRegex, repositoryUrlRegex };
+
+exports.getPluginByArguments = (userInput) => {
+  const args = parseArguments(userInput);
+  const isValidPluginName = pluginNameRegex.test(args['repository-name']);
+  const isValidRepositoryUrl = repositoryUrlRegex.test(args['repository-url']);
+
+  if (!isValidPluginName || !isValidRepositoryUrl) {
+    console.log(`\n${chalk.red('✘')} Repository name and/or url invalid.`);
+    return customPlugin;
```
Comment (k-y0u) [inline src/services/helper.js:85]: jump line before

File: src/services/helper.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  customPlugin.name = args['repository-name'];
+  customPlugin.url = args['repository-url'];
+  return customPlugin;
```
Comment (k-y0u) [inline src/services/helper.js:90]: jump line before

File: src/services/helper.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  if (installationType === 'official') {
+    const officialPlugins = await getOfficialPlugins();
+
+    if (officialPlugins.length === 0) {
+      return [];
+    }
+
+    return officialPlugins.map((plugin) => ({ name: plugin.name, url: plugin.url }));
+  }
+  const customPromptPlugin = await getCustomPlugin();
```
Comment (k-y0u) [inline src/services/helper.js:105]: jump line before

File: src/services/helper.js [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+  if (!customPromptPlugin.name || !customPromptPlugin.url) {
+    console.log(`\n${chalk.red('✘')} No plugin has been installed.`);
+    return [];
```
Comment (k-y0u) [inline src/services/helper.js:109]: jump line before

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: src/services/helper.js (hunk 2 above)
Comment (Zorin95670) [reply src/services/helper.js]: No more relevant, related to old code version.

File: src/services/helper.js (hunk 3 above)
Comment (Zorin95670) [reply src/services/helper.js]: No more relevant, related to old code version.

File: src/services/helper.js (hunk 4 above)
Comment (Zorin95670) [reply src/services/helper.js]: No more relevant, related to old code version.

File: src/services/helper.js (hunk 5 above)
Comment (Zorin95670) [reply src/services/helper.js]: No more relevant, related to old code version.

File: src/services/helper.js (hunk 8 above)
Comment (Zorin95670) [reply src/services/helper.js]: No more relevant, related to old code version.

File: src/services/helper.js (hunk 7 above)
Comment (Zorin95670) [reply src/services/helper.js]: No more relevant, related to old code version.

File: src/services/helper.js (hunk 11 above)
Comment (Zorin95670) [reply src/services/helper.js]: No more relevant, related to old code version.

File: src/services/helper.js (hunk 12 above)
Comment (Zorin95670) [reply src/services/helper.js]: No more relevant, related to old code version.

File: README.md [hunk 13]
```diff
@@ -48,6 +48,25 @@ Add this to package.json scripts:
     "plugin:init": "node node_modules/leto-modelizer-plugin-cli init"
 }
 ```
+### List of official plugins
```
Comment (Zorin95670) [inline README.md:51]: How to define official plugins

Comment (Zorin95670) [reply README.md:51]: Indicate before that we must define an official plugins list if we want to use this CLI

File: README.md [hunk 14]
```diff
@@ -48,6 +48,25 @@ Add this to package.json scripts:
...
+
+A list of official plugins to install can be defined. For that, create a `cli.config.json` file at the root of Leto Modelizer project. Each object represent a plugin. Example :
```
Comment (Zorin95670) [inline README.md:53]: Reword your sentence to be more like an "how to"

File: changelog.md [hunk 15]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
 
+## Unreleased
+
+### Added
+
+- Add prompt to choose between official or custom plugins via `install` command.
```
Comment (Zorin95670) [inline changelog.md:12]: Add:
```
- Externalize official plugins list in leto-modelizer
```

Comment (justinehell) [reply changelog.md:12]: done

File: src/services/PluginValidator.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    this.name = name || null;
+    this.url = url || null;
+  }
+
+  /**
+   * Return whether the plugin name and the repository URL are valid.
+   * @returns {boolean} True if the plugin name and the repository URL are valid, otherwise false.
+   */
+  isValid() {
+    return pluginNameRegex.test(this.name) && repositoryUrlRegex.test(this.url);
```
Comment (Zorin95670) [inline src/services/PluginValidator.js:24]:
```
return this.isPluginNameValid() && this.isRepositoryUrlValid();
```

File: README.md (hunk 13 above)
Comment (justinehell) [reply README.md]: first done, second -> add log to indicate `no official plugins available` if official plugins list is empty or doesn't exist

File: src/services/PromptPluginRetriever.js [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * Read the official plugins from the `leto-modelizer-plugin-cli.json` file, call the corresponding
+ * function to prompt the user to choose official plugin(s) to install.
+ *
+ * @returns {Promise<Array>} An array of objects representing the plugin(s),
+ * otherwise an empty array.
+ */
+async function getOfficialPlugins() {
+  if (!fs.existsSync('leto-modelizer-plugin-cli.json')) {
```
Comment (Zorin95670) [inline src/services/PromptPluginRetriever.js:74]: change by:
```js
function getPluginsFromConfig() {
  if (!fs.existsSync('leto-modelizer-plugin-cli.json')) {
    return [];
  }
  
  return JSON.parse(fs.readFileSync('leto-modelizer-plugin-cli.json', 'utf8'));
}

async function getOfficialPlugins() {
  const plugins = getPluginsFromConfig();

  if (plugins.length === 0) {
    console.log(`\n${chalk.red('✘')} No official plugins available.`);

    return [];
  }
(...)  
```

Comment (justinehell) [reply src/services/PromptPluginRetriever.js:73]: done

**Pull request merged**